### DISCUSSION TITLE
PI Zero W 2 - Additional Instructions

### DISCUSSION THREAD

**Discussion by KJ7PPK** [Nachtzuster/BirdNET-Pi#258] [General]
Hey everyone, I wrote the latest 64-bit RPi OS LITE image on a 32GB microSD card with the official RPi Imager application to use in my Pi Zero 2 W and there were a few things not noted that I had to figure out for myself...

1. In the existing reference to Pi Zero 2 W installation, the provided 3 commands to expand swap space and turn off the Wi-Fi power saving function will error out on a headless Pi Zero 2 W installation as the /etc/rc.local file will not exist. 
2. I had to create this file with "sudo nano /etc/rc.local" and populate with the following lines: 
#!/bin/sh -e
COMMANDS
exit 0
3. Set permissions on that file using "sudo chmod u+x /etc/rc.local" and then run the third line provided in the guidance.

During installation from Nachtzuster's repo, I kept getting an "OS error out of space" message. This was due to pip seeing that /tmp was out of space. I was able to increase this by running: "sudo mount -o remount,size=5G /tmp/"

After this, I had to run the script from the repo (/scripts/uninstall.sh). Then, "sudo rm -r /home/USERNAME/BirdNET-Pi". 

After all of this, I was able to run the one-liner bash command and successfully install the application.

**Final note: Despite Alsamixer showing the correct default device, I could not get my USB sound card input to work in the application. I started from scratch with the sound card installed and it worked out fine.**

Another note.... the performance is less than optimal. My recommendation is to run BirdnetCAGE in a docker container on a capable machine, then setup ffmpeg audio streams with your cheap pi zero hardware streaming to said server. The Zero devices will just function as a collection means and the server can do all the heavy lifting. I promised my lady a functional BirdnetPI setup this weekend and I'm currently doing a 24TB disk clone on the server, so this will do for now. Great way to dip your toes in.

**Comment by heyho74** [comment]
Thanks for this, I posted on the old page but have had no responses, but this is the information I've been looking for.

When you say you started from scratch to get the soundcard working, do you mean just the birdnet software, or Pi OS?

**Reply by DSRZH** [reply]
thanks @brokensolenoid for your tipps. I also could apply them. birdNET is now up and running on a  Pi Zero 2 W.
I'm not deep into programming so  did't understand how to (/scripts/uninstall.sh), but it works also without uninstall.
PS: a half year ago these steps where not necessary. at this time i made my last installation on another Pi Zero 2 W.

**Reply by Award1000** [reply]
Thank you for this. That change on remounting the tmp directory worked. No need to reinstall. Just re-ran the install script.

**Reply by DavidGrim** [reply]
Did you just re-run ~/BirdNET-Pi/scripts $ ./install_birdnet.sh ?

**Comment by Grey-Lancaster** [comment]
it appears that you have to let an install fail before you can do this
"sudo mount -o remount,size=5G /tmp/"


Is there a better way?
Thanks,
Grey

**Comment by dalogue1** [comment]
Your description for the Pi Zero 2W is perfect!! 

I did have to set the mount command to 10G to get it to work i.e.  "sudo mount -o remount,size=10G /tmp/"

The only thing that would make it better was if there was a script that would do all those changes automatically prior to running the installation.

**Comment by dalogue1** [comment]
Also, a corrected version of the Pi Zero 2 W instructions for the Nachtzuster Github page is probably in order. We could start with the existing page from https://github.com/mcguirepr89/BirdNET-Pi/wiki/RPi0W2-Installation-Guide and correct the problems and maybe add the suggestions from above. The current modification of the rc.local file is very confusing to a new user. Possibly, just using the nano editor to begin with makes more sense. Using nano to edit the dphys-swapfile would make sense as well.

Possibly, the standard BirdNET-Pi installation script could be modified to mount the /tmp/ location with 5G or more for the installation.

**Of course, the best solution would be for all of the required changes to be made to the standard installation so that the Pi Zero 2 W works by default.  That is for someone else to determine. In that case, no special page is required.**

A Pi Zero 2 W is a cost effective choice for new users as well as gifts for friends and family.  (I have a prepared list on Amazon with a selection of everything needed from a single source.  Not that Jeff Bezos needs more money though!)

**Comment by riggo37** [comment]
I have found that installing birdnet-go https://github.com/tphakala/birdnet-go in a docker container on a Pi Zero 2W running Dietpi https://dietpi.com/  a much more stable set up.

**Comment by gpbmike** [comment]
Thanks for these steps. I had to go through them twice (uninstall/install twice) for it to work for me. Cheers.

**Comment by dalogue1** [comment]
I have found that the Pi Zero-2W has the possibility of overheating especially if you enable some overlap.  

I have changed my Pi Zero 2W systems to have a nice case that has a built in heat sink.  https://www.amazon.com/dp/B08837L144?ref=ppx_yo2ov_dt_b_fed_asin_title

Under load it drops the CPU temperature by about 10 C.  (You can see the CPU Temperature in the System Info under utilities.)

**Comment by hpk60** [comment]
I installed birdnet-pi today on a brandnew Zero 2WH with a SanDisk Extreme Pro 32GB  without any issues or modifications. 
First installed with Raspberry Pi Imager v1.9.4 Raspberry Pi OS Lite (64-bit) then apt-get update and full-upgrade.
As third step birdnet-pi with 
`curl -s https://raw.githubusercontent.com/Nachtzuster/BirdNET-Pi/main/newinstaller.sh | bash`
So the installation worked fine. Now starts the testing :-)

Does anyone have recommendations for best settings for Raspi Zero 2WH or can they be the same as for Raspi 4B or Raspi 5?

**Reply by dalogue1** [reply]
**1. Enlarge the swap file size**

`sudo nano /etc/dphys-swapfile`

Make the following updates and exit the editor :
`CONF_SWAPSIZE=2048`
`CONF_MAXSWAP=4096`

**2. Enlarge the /tmp directory size**

`sudo nano /etc/fstab`

Add the following line to the end and exit the editor :
`tmpfs     /tmp     tmpfs     defaults,size=10G,mode=1777     0     0`

**3. Reboot to implement the changes and run the curl install script for BirdNET-Pi**

**4. Enable the watchdog to reboot the Zero 2W if it stops after a random period of time.**

`sudo nano /etc/systemd/system.conf`

Make the following updates and exit the editor:
`RuntimeWatchdogSec=15`
`RebootWatchdogSec=2min`

Run the following command or reboot:
`sudo systemctl daemon-reload`

This article describes the watchdog built into the pi:  https://pimylifeup.com/raspberry-pi-watchdog/
Whether you use the standard system method or watchdog application does not matter. 

**5. Disable Wi-Fi power-saving mode:**

Run the command
`sudo crontab -e `   

Add the following line:
`* * * * * /sbin/iwconfig wlan0 power off`

This is a simplification of the following article:
https://unix.stackexchange.com/questions/269661/how-to-turn-off-wireless-power-management-permanently

**Reply by cdzombak** [reply]
Particularly if you're using swap, IMO you absolutely _must_ use a high-endurance or industrial-endurance SD card. Normal SD cards will often die in a year or less used in a Raspberry Pi with swap.

Personally, I recommend the SanDisk MAX Endurance cards. Be sure to buy from a trusted seller and avoid counterfeits.

**Reply by hpk60** [reply]
Thank you, I've added above configurations.

@cdzombak: I'll use the Pi 0W not for longtime analisys, I'll leave it for a day or a night to get some informations about birds in the wild. 
At home, I have a RPI4B that runs like a charm.

**Comment by DavidGrim** [comment]
Replacing regular brand name SD cards with a SanDisk Max Endurance resolved the hang errors I saw in just a day or two with the regular cards.

**Comment by dalogue1** [comment]
Has anyone **recently** performed a fresh install of BirdNET-Pi on a Pi Zero 2W and had it work?

I tried to load a fresh version of Raspberry PI OS Bookworm 64-bit Lite and BirdNET-Pi on a Pi Zero 2W and I get the "Caddy Works" webpage instead of the BirdNET-Pi webpage.  I retried 5 times with no success.  Instead, I finally copied an old instance of a working BirdNET-Pi on a Pi Zero 2W and just updated BirdNET-Pi to the current version.

I have in the past loaded the Pi Zero 2W many times successfully, but have recently started using RPi 5 for reliability. (Often a Pi Zero 2W hangs after a few days of operation for no explainable reason. This does not happen with a RPi 5.)  Performing the same fresh install on a RPi 5 works just fine.

**Reply by mydevices514** [reply]
I just got it working yesterday! 
 - Pi Zero 2 W with 64gb SD

After many failed attempts, mostly due to 'not enough space' errors, I went back to Pi Imager 1.9.4, and used everyone's input. It took a couple of tries, as 5G was not enough.

After it failed, I did dalogue1's list of instructions (1 through 5) 

Then I removed the previous attempt,
 _sudo rm -r /home/USERNAME/BirdNET-Pi_
made tmp 10G 
 _sudo mount -o remount,size=**10**G /tmp/_
then reran the install 
 _curl -s https://raw.githubusercontent.com/Nachtzuster/BirdNET-Pi/main/newinstaller.sh | bash_
 
 Still testing, I already found that my USB extension is impeding my mic, but without it I got it to recognize a few birds from a video.

**Reply by dalogue1** [reply]
I modified step 2 to 10G if that helps.

Well you got further than I did since you got a BirdNET-Pi web page to show instead of "Caddy Works!"

**Comment by dkerlee** [comment]
I like the suggestion above about installing BirdNET on a more capable machine (I have one) via Docker, then piping the audio to that server via ffmpeg and a raspberrypi zero. I think I'm going to give that a try! I'll report back.

**Comment by DavidGrim** [comment]
I have two Raspberry Pi Zero 2 birdnet devices and I occasionally, once per week or more, see that one or the other has hanged. The older one running longer with 100,000+ detections might hang more often.  I have not seen my Raspberry Pi 4 or 5 birdnet Pi devices hang. I did change the swap space as instructed and use SanDisk max endurance SD cards.

**Reply by dalogue1** [reply]
I finally got fed up with the Pi Zero 2Ws and replaced them with RPi 5s.  I did my level best but they just have something that hangs while the RPi 5 does not.

**Reply by DavidGrim** [reply]
For several months, the original zero seemed fairly reliable. It still hanged occasionally, but the frequency seems to have increased.  I cleaned up the dB, removing detections older than a month from 100,000+ to 8000 detections. The birdnet browser ui states the total decreased to 8000 but the birds.db size remains at 26,017,792. My newer zero has a similar number of detections but its  birds.db size is only 1,794,048.

I also removed  the original one's case and added a timer job to measure and log temperature hourly to see that data after a crash in case there is a thermal issue.

**Reply by dalogue1** [reply]
I have noticed that sometimes the entire device is hung and at other times it has lost the network connection permanently instead.

You can see the difference based upon continued identifications in the local database but no BirdWeather reports.

**For the network hang problem**, 
I do not have a keyboard and display outside to attempt to just restart the networking.  Maybe a cron job that would restart the networking every few hours might help.  Also, for one of my Pi Zero 2Ws I replaced the internal antenna with an external antenna by soldering on a  connector.
<img width="169" height="130" alt="image" src="https://github.com/user-attachments/assets/3a679509-b136-42d3-9e79-f11474282d07" />
It had a much stronger WiFi signal but still occasionally becomes disconnected.  The best solution appears to be to have the Pi Zero inside of the house with the microphone wire being routed out of a window with a long microphone wire.

**For the total hang problem,** 
The Pi Zero 2W does have a hardware watchdog that is independent of the normal CPU calculations.  I did try this as described above on June 25, 2025 but this did not eliminate the problem.

_Maybe a future Pi Zero with more memory and more solid WiFi will work fine, but for now the RPi5 is worth the extra cost to me._ 
(_Adding to that the reliability and speed of an nvme drive makes it an easier decision as well._)

**Reply by DavidGrim** [reply]
> For several months, the original zero seemed fairly reliable. It still hanged occasionally, but the frequency seems to have increased. I cleaned up the dB, removing detections older than a month from 100,000+ to 8000 detections. The birdnet browser ui states the total decreased to 8000 but the birds.db size remains at 26,017,792. My newer zero has a similar number of detections but its birds.db size is only 1,794,048.
> 
> I also removed the original one's case and added a timer job to measure and log temperature hourly to see that data after a crash in case there is a thermal issue.

Even after cleaning up the database (deleting old detections and vacuum the db), making sure I have 2A (10W) power supplies and max endurance sd cards, adding a watchdog timer (and then removing it), I still see either frequent (at least once per day) reboots or hangs with my pi 02's running birdnet. I have not seen any reboots or hangs with Pi4's or Pi5's. The hangs either occur with the power LED on or off. A few times, I have seen the system at a command line showing 127.0.0.1, i.e. it lost the network.

Temperature is always 47-51C monitored hourly and not correlated to any hang/reboot.
